service/index/index_integration_test.go

```diff
@@ -0,0 +1,238 @@
+
+import (
+	"sort"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/onflow/flow-go/model/flow"
+
+	"github.com/optakt/flow-dps/codec/zbor"
+	"github.com/optakt/flow-dps/service/index"
+	"github.com/optakt/flow-dps/service/storage"
+	"github.com/optakt/flow-dps/testing/helpers"
+	"github.com/optakt/flow-dps/testing/mocks"
+)
+
+func TestIndex(t *testing.T) {
+
+	t.Run("first", func(t *testing.T) {
+		t.Parallel()
+
+		reader, writer := setupIndex(t)
+
+		assert.NoError(t, writer.First(mocks.GenericHeight))
+		// Close the writer to make it commit its transactions.
+		assert.NoError(t, writer.Close())
+
+		got, err := reader.First()
+
+		assert.NoError(t, err)
+		assert.Equal(t, mocks.GenericHeight, got)
+	})
+
+	t.Run("last", func(t *testing.T) {
+		t.Parallel()
+
+		reader, writer := setupIndex(t)
+
+		assert.NoError(t, writer.Last(mocks.GenericHeight))
+		// Close the writer to make it commit its transactions.
+		assert.NoError(t, writer.Close())
+
+		got, err := reader.Last()
+
+		assert.NoError(t, err)
+		assert.Equal(t, mocks.GenericHeight, got)
+	})
+
+	t.Run("height", func(t *testing.T) {
+		t.Parallel()
+
+		reader, writer := setupIndex(t)
+
+		assert.NoError(t, writer.Height(mocks.GenericIdentifier(0), mocks.GenericHeight))
+		// Close the writer to make it commit its transactions.
+		assert.NoError(t, writer.Close())
+
+		got, err := reader.HeightForBlock(mocks.GenericIdentifier(0))
+
+		assert.NoError(t, err)
+		assert.Equal(t, mocks.GenericHeight, got)
+	})
+
+	t.Run("commit", func(t *testing.T) {
+		t.Parallel()
+
+		reader, writer := setupIndex(t)
+
+		assert.NoError(t, writer.Commit(mocks.GenericHeight, mocks.GenericCommit(0)))
+		// Close the writer to make it commit its transactions.
+		assert.NoError(t, writer.Close())
+
+		got, err := reader.Commit(mocks.GenericHeight)
+
+		assert.NoError(t, err)
+		assert.Equal(t, mocks.GenericCommit(0), got)
+	})
+
+	t.Run("header", func(t *testing.T) {
+		t.Parallel()
+
+		reader, writer := setupIndex(t)
+
+		assert.NoError(t, writer.Header(mocks.GenericHeight, mocks.GenericHeader))
+		// Close the writer to make it commit its transactions.
+		assert.NoError(t, writer.Close())
+
+		got, err := reader.Header(mocks.GenericHeight)
+
+		assert.NoError(t, err)
+		assert.Equal(t, mocks.GenericHeader, got)
+	})
+
+	t.Run("payloads", func(t *testing.T) {
+		t.Parallel()
+
+		reader, writer := setupIndex(t)
+
+		paths := mocks.GenericLedgerPaths(4)
+		payloads := mocks.GenericLedgerPayloads(4)
+
+		assert.NoError(t, writer.First(mocks.GenericHeight))
+		assert.NoError(t, writer.Last(mocks.GenericHeight))
+		assert.NoError(t, writer.Payloads(mocks.GenericHeight, paths, payloads))
+		// Close the writer to make it commit its transactions.
+		assert.NoError(t, writer.Close())
+
+		got, err := reader.Values(mocks.GenericHeight, paths)
+
+		assert.NoError(t, err)
+		assert.Equal(t, mocks.GenericLedgerValues(4), got)
```

Comment:
I'm surprised this passes. Shouldn't there be an ordering mismatch sometimes? I would intuitively think that it should be `assert.EqualElements` (or what's the name)?